### ISSUE TITLE
TCP proxying not working as expected

### ISSUE THREAD

**Issue by rudimk** [mholt/caddy-l4#25]
Hi Matt!

First off - love Caddy and its premise!

I'm running Caddy v2.4.3 (`h1:Y1FaV2N4WO3rBqxSYA8UZsZTQdN+PwcoOcAiZTM8C0I=`) on Ubuntu 18.04 after doing a custom build on the website, making sure I added `caddy2-proxyprotocol` and `caddy-l4` as addon features. Here the basic idea is to have a simply proxy listening on port 3306, for MySQL on another server. Here's my JSON config:

```
{
   "apps":{
      "layer4":{
         "servers":{
            "mysql":{
               "listen":[
                  ":3306"
               ],
               "routes":[
                  {
                     "handle":[
                        {
                           "handler":"proxy",
                           "proxy_protocol":"v2",
                           "upstreams":[
                              {
                                 "dial":[
                                    "tcp/10.0.0.3:3306"
                                 ]
                              }
                           ]
                        }
                     ]
                  }
               ]
            }
         }
      }
   }
}
```
The MySQL server is reachable from the server running caddy, so the network isn't an issue. But running telnet produces an odd error:

```
$ telnet 127.0.0.1 3306
Connected to localhost.
Escape character is '^]'.
[
5.7.34-0ubuntu0.18.04.1E"n !b,dMj	HD3=mysql_native_password!#08S01Got packets out of order^CConnection closed by foreign host.
```

So it's able to proxy requests to the MySQL server - but something's going awry with the packets. Trying to access this from a MySQL client throws the following error:

```
Lost connection to MySQL server at 'reading authorization packet', system error: 0
```

I went through MySQL's documentation, thinking maybe there's something up with MySQL - although this hasn't happened when I try accessing the MySQL server directly. Anyway, I tried tweaking connection timeouts, but to no avail. I then tried running the same configuration against a Redis instance, substituting the correct IPs and ports in the upstream - it didn't work, `redis-cli` refused to connect to Caddy.

At this point I'm fairly certain it's definitely something up with the configuration I've defined - I adapted it from the exampes in the `README`. Would love some tips :)

EDIT - I've also tried various combinations of using/not using the `tcp/` prefixes in the listener and the handlers - I'm afraid it doesn't make a difference.

**Comment by francislavoie** [comment]
Are you sure MySQL supports the PROXY protocol? That's typically only supported by HTTP servers as far as I know.

**Comment by rudimk** [comment]
A good question. Actually it does, and HAProxy allows for proxying to MySQL using the PROXY protocol. At any rate, I removed the `proxy_protocol` directive, and this time, the error changes slightly. 

Over telnet, the connection is immediately closed:
```
> telnet localhost 3306
Trying ::1...
Connected to localhost.
Escape character is '^]'.
Connection closed by foreign host.
```

Over a proper MySQL client, the error message changes slightly: `MySQL said: Lost connection to MySQL server at 'reading initial communication packet', system error: 0`

**Comment by rudimk** [comment]
....and it worked! Turns out MySQL wasn't running, and that's something I overlooked whilst trying to proxy without the `proxy_protocol` parameter. 

Another thing I'd like to add - whilst MySQL does support the PROXY protocol, one needs to add `proxy_protocol_networks` to the `mysqld.cnf` file; it's used to specify IPs or CIDRs from which PROXY requests can be accepted. So if one needs to use the `proxy_protocol` parameter to ensure their source IPs aren't lost in transit, they'd have to reconfigure MySQL too. 

Closing this issue since this is sorted. Thanks a ton, @francislavoie!
